File: 2020/day12/sol.cpp
#include <iostream>
#include <fstream>

int main(){
    std::ifstream input("input");

    int x = 0, y = 0;
    int wx = 10, wy = 1;
    int ship_x = 0, ship_y = 0;
    int direction = 90;

    std::string line;
    while (std::getline(input, line)) {
        char action = line[0];
        int units = std::stoi(line.substr(1));

        switch (action) {
            case 'N':
                y += units;
                wy += units;
                break;
            case 'S':
                y -= units;
                wy -= units;
                break;
            case 'E':
                x += units;
                wx += units;
                break;
            case 'W':
                x -= units;
                wx -= units;
                break;
            case 'L': {
                for (int j = units / 90; j > 0; j--) {
                    std::swap(wx, wy);
                    wx = -wx;
                }
                direction -= units;
                break;
            }
            case 'R': {
                for (int j = units / 90; j > 0; j--) {
                    std::swap(wx, wy);
                    wy = -wy;
                }
                direction += units;
                break;
            }
            case 'F':
                switch (direction) {
                    case 0:   y += units; break;
                    case 90:  x += units; break;
                    case 180: y -= units; break;
                    case 270: x -= units; break;
                }
                ship_x += units * wx;
                ship_y += units * wy;
                break;
        }

        if      (direction >= 360) direction -= 360;
        else if (direction < 0)    direction += 360;
    }

    std::cout << "part 1: " << abs(x) + abs(y) << '\n';
    std::cout << "part 2: " << abs(ship_x) + abs(ship_y) << '\n';

    return 0;
}

File: 2020/day12/sol.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("input").expect("Can't read puzzle input file.");

    let instructions: Vec<_> = input
        .lines()
        .map(|line| {
            let (action, value) = line.split_at(1);
            let action = action.chars().next().unwrap();
            let value = value.parse::<i32>().unwrap();
            (action, value)
        })
        .collect();

    let mut north = 0;
    let mut east = 0;
    let mut direction = 90;

    instructions
        .iter()
        .for_each(|&(action, value)| match action {
            'N' => north += value,
            'S' => north -= value,
            'E' => east += value,
            'W' => east -= value,
            'L' => {
                direction += 360 - value;
                direction %= 360;
            }
            'R' => {
                direction += value;
                direction %= 360;
            }
            'F' => match direction {
                0 => north += value,
                90 => east += value,
                180 => north -= value,
                270 => east -= value,
                _ => {}
            },
            _ => {}
        });

    println!("part 1: {}", north.abs() + east.abs());

    let mut waypoint_x = 10;
    let mut waypoint_y = 1;
    let mut ship_x = 0;
    let mut ship_y = 0;

    instructions
        .iter()
        .for_each(|&(action, value)| match action {
            'N' => waypoint_y += value,
            'S' => waypoint_y -= value,
            'E' => waypoint_x += value,
            'W' => waypoint_x -= value,
            'L' => {
                for _ in 0..value / 90 {
                    core::mem::swap(&mut waypoint_x, &mut waypoint_y);
                    waypoint_x = -waypoint_x;
                }
            }
            'R' => {
                for _ in 0..value / 90 {
                    core::mem::swap(&mut waypoint_x, &mut waypoint_y);
                    waypoint_y = -waypoint_y;
                }
            }
            'F' => {
                ship_x += value * waypoint_x;
                ship_y += value * waypoint_y;
            }
            _ => {}
        });

    println!("part 2: {}", ship_x.abs() + ship_y.abs());
}
